### DISCUSSION TITLE
Serialization of enums

### DISCUSSION THREAD

**Discussion by lukehutch** [serverpod/serverpod#715] [Q&A] [answered]
Serverpod currently supports serialization of "all serializable types", but as far as I can tell, enums are not serializable out of the box. Is that true? If so, what would it take for Serverpod to add enum serialization and deserialization?

**Comment by lukehutch** [comment]
Maybe `jsonize` could help? It handles enums, DateTime, etc.

https://pub.dev/packages/jsonize

**Comment by fischerscode** [comment, marked as answer]
While this is currently not documented, you can define serializable enums just like classes. https://github.com/serverpod/serverpod/blob/eec8af12e41aef94cc9f162358d9bb3e77bd9eef/tests/serverpod_test_server/lib/src/protocol/test_enum.yaml#L1-L12

**Reply by lukehutch** [reply]
Awesome, thanks for the pointer!

**Reply by lukehutch** [reply]
I assume these work just fine with database serialization too? Does this store integers or strings (enum constant names) in the database?

**Reply by fischerscode** [reply]
It uses the index (integer). So you must not delete a value from the yaml file (everything shifts).

**Reply by lukehutch** [reply]
It would be great to add an option to use the enum constant itself, rather than the index, to protect against this...

**Reply by fischerscode** [reply]
I'll probably provide a PR that adds support for user/developer defined values in the future. For now, it should be safe to use the index since you could run a custom SQL that shifts the indexes.

**Reply by andriyndev** [reply]
I think enum's values should never be serialized by numeric values because if the app's developer edits the enum so that the values of the current elements are shifted, it'll cause nasty consequences. For example, user selects one value, but it's written to the database as another one, and the worse is that they'll see it as they specified it while other users with more recent versions of the app will see it as it was written to the database. It might be especially dangerous if the data represents delicate information like user's religion, political view, sexual orientation, gender etc. And even if the app's developers are aware of this nuance, they still might be susceptible to human factor. And if the mistake will be made, it might be impossible to fix the consequences without interacting with each user who triggered writes of the data to the database because there is no way for the app's developers to know whether the value is what the user actually selected.

**Comment by vlidholt** [comment]
I added the documentation tag, so that we add this to the documentation!
